doc/libpmem.3
.TH "libpmem" "3" "" "" ""
.SS NAME
.PP
\f[B]libpmem\f[] − persistent memory support library
.SS SYNOPSIS
.IP
.nf
\f[C]
#include\ <libpmem.h>
cc\ ...\ \-lpmem
\f[]
.fi
.SS Most commonly used functions:
.IP
.nf
\f[C]
int\ pmem_is_pmem(const\ void\ *addr,\ size_t\ len);

void\ pmem_persist(const\ void\ *addr,\ size_t\ len);

int\ pmem_msync(const\ void\ *addr,\ size_t\ len);

void\ *pmem_map_file(
\ \ \ \ const\ char\ *path,\ size_t\ len,
\ \ \ \ int\ flags,\ mode_t\ mode,
\ \ \ \ size_t\ *mapped_lenp,\ int\ *is_pmemp);

int\ pmem_unmap(void\ *addr,\ size_t\ len);
\f[]
.fi
.SS Partial flushing operations:
.IP
.nf
\f[C]
void\ pmem_flush(const\ void\ *addr,\ size_t\ len);

void\ pmem_drain(void);

int\ pmem_has_hw_drain(void);
\f[]
.fi
.SS Copying to persistent memory:
.IP
.nf
\f[C]
void\ *pmem_memmove_persist(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len);

void\ *pmem_memcpy_persist(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len);

void\ *pmem_memset_persist(void\ *pmemdest,\ int\ c,\ size_t\ len);

void\ *pmem_memmove_nodrain(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len);

void\ *pmem_memcpy_nodrain(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len);

void\ *pmem_memset_nodrain(void\ *pmemdest,\ int\ c,\ size_t\ len);
\f[]
.fi
.SS Library API versioning:
.IP
.nf
\f[C]
const\ char\ *pmem_check_version(unsigned\ major_required,\ unsigned\ minor_required);
\f[]
.fi
.SS Error handling:
.IP
.nf
\f[C]
const\ char\ *pmem_errormsg(void);
\f[]
.fi
.SS DESCRIPTION
.PP
\f[B]libpmem\f[] provides low\-level \f[I]persistent memory\f[] (pmem)
support for applications using direct access storage (DAX), which is
storage that supports load/store access without paging blocks from a
block storage device.
Some types of \f[I]non\-volatile memory DIMMs\f[] (NVDIMMs) provide this
type of byte addressable access to storage.
A \f[I]persistent memory aware file system\f[] is typically used to
expose the direct access to applications.
Memory mapping a file from this type of file system results in the
load/store, non\-paged access to pmem.
.PP
This library is for applications that use persistent memory directly,
without the help of any library\-supplied transactions or memory
allocation.
Higher\-level libraries that build on \f[B]libpmem\f[] are available and
are recommended for most applications, see:
.IP \[bu] 2
\f[B]libpmemobj\f[](3), a general use persistent memory API, providing
memory allocation and transactional operations on variable\-sized
objects.
.IP \[bu] 2
\f[B]libpmemblk\f[](3), providing pmem\-resident arrays of fixed\-sized
blocks with atomic updates.
.IP \[bu] 2
\f[B]libpmemlog\f[](3), providing a pmem\-resident log file.
.PP
Under normal usage, \f[B]libpmem\f[] will never print messages or
intentionally cause the process to exit.
The only exception to this is the debugging information, when enabled,
as described under \f[B]DEBUGGING AND ERROR HANDLING\f[] below.
.SS MOST COMMONLY USED FUNCTIONS
.PP
Most pmem\-aware applications will take advantage of higher level
libraries that alleviate the application from calling into
\f[B]libpmem\f[] directly.
Application developers that wish to access raw memory mapped persistence
directly (via \f[B]mmap\f[](2)) and that wish to take on the
responsibility for flushing stores to persistence will find the
functions described in this section to be the most commonly used.
.IP
.nf
\f[C]
int\ pmem_is_pmem(const\ void\ *addr,\ size_t\ len);
\f[]
.fi
.PP
The \f[C]pmem_is_pmem()\f[] function returns true only if the entire
range [\f[C]addr\f[], \f[C]addr\f[]+\f[C]len\f[]) consists of persistent
memory.
A true return from \f[C]pmem_is_pmem()\f[] means it is safe to use
\f[C]pmem_persist()\f[] and the related functions below to make changes
durable for that memory range.
.PP
The implementation of \f[C]pmem_is_pmem()\f[] requires a non\-trivial
amount of work to determine if the given range is entirely persistent
memory.
For this reason, it is better to call \f[C]pmem_is_pmem()\f[] once when
a range of memory is first encountered, save the result, and use the
saved result to determine whether \f[C]pmem_persist()\f[] or
\f[B]msync\f[](2) is appropriate for flushing changes to persistence.
Calling \f[C]pmem_is_pmem()\f[] each time changes are flushed to
persistence will not perform well.
.RS
.PP
WARNING: Using \f[C]pmem_persist()\f[] on a range where
\f[C]pmem_is_pmem()\f[] returns false may not do anything useful \[en]
use \f[B]msync\f[](2) instead.
.RE
.IP
.nf
\f[C]
void\ pmem_persist(const\ void\ *addr,\ size_t\ len);
\f[]
.fi
.PP
Force any changes in the range
[\f[C]addr\f[],\f[C]addr\f[]+\f[C]len\f[]) to be stored durably in
persistent memory.
This is equivalent to calling \f[B]msync\f[](2) but may be more optimal
and will avoid calling into the kernel if possible.
There are no alignment restrictions on the range described by
\f[C]addr\f[] and \f[C]len\f[], but \f[C]pmem_persist()\f[] may expand
the range as necessary to meet platform alignment requirements.
.RS
.PP
WARNING: Like \f[B]msync\f[](2), there is nothing atomic or
transactional about this call.
Any unwritten stores in the given range will be written, but some stores
may have already been written by virtue of normal cache
eviction/replacement policies.
Correctly written code must not depend on stores waiting until
\f[C]pmem_persist()\f[] is called to become persistent \[en] they can
become persistent at any time before \f[C]pmem_persist()\f[] is called.
.RE
.IP
.nf
\f[C]
int\ pmem_msync(const\ void\ *addr,\ size_t\ len);
\f[]
.fi
.PP
The function \f[C]pmem_msync\f[]() is like \f[C]pmem_persist()\f[] in
that it forces any changes in the range [\f[C]addr\f[],
\f[C]addr\f[]+\f[C]len\f[]) to be stored durably.
Since it calls \f[C]msync()\f[], this function works on either
persistent memory or a memory mapped file on traditional storage.
\f[C]pmem_msync()\f[] takes steps to ensure the alignment of addresses
and lengths passed to \f[C]msync()\f[] meet the requirements of that
system call.
It calls \f[C]msync()\f[] with the \f[C]MS_SYNC\f[] flag as described in
\f[B]msync\f[](2).
Typically the application only checks for the existence of persistent
memory once, and then uses that result throughout the program, for
example:
.IP
.nf
\f[C]
/*\ do\ this\ call\ once,\ after\ the\ pmem\ is\ memory\ mapped\ */
int\ is_pmem\ =\ pmem_is_pmem(rangeaddr,\ rangelen);

/*\ \&...\ make\ changes\ to\ a\ range\ of\ pmem\ \&...\ */

/*\ make\ the\ changes\ durable\ */
if\ (is_pmem)
\ \ pmem_persist(subrangeaddr,\ subrangelen);
else
\ \ pmem_msync(subrangeaddr,\ subrangelen);

/*\ \&...\ */
\f[]
.fi
.PP
The return value of \f[C]pmem_msync()\f[] is the return value of
\f[B]msync\f[](), which can return \-1 and set \f[C]errno\f[] to
indicate an error.
.IP
.nf
\f[C]
void\ *pmem_map_file(
\ \ \ \ const\ char\ *path,\ size_t\ len,int\ flags,
\ \ \ \ mode_t\ mode,\ size_t\ *mapped_lenp,\ int\ *is_pmemp);
\f[]
.fi
.PP
Given a \f[C]path\f[], \f[C]pmem_map_file()\f[] function creates a new
read/write mapping for the named file.
It will map the file using \f[B]mmap\f[](2), but it also takes extra
steps to make large page mappings more likely.
.PP
On success, \f[C]pmem_map_file()\f[] returns a pointer to mapped area.
If \f[C]mapped_lenp\f[] is not NULL, the length of the mapping is also
stored at the address it points to.
The \f[C]is_pmemp\f[] argument, if non\-NULL, points to a flag that
\f[C]pmem_is_pmem()\f[] sets to say if the mapped file is actual pmem,
or if \f[C]msync()\f[] must be used to flush writes for the mapped
range.
On error, \f[C]NULL\f[] is returned, \f[C]errno\f[] is set
appropriately, and \f[C]mapped_lenp\f[] and \f[C]is_pmemp\f[] are left
untouched.
.PP
The \f[C]flags\f[] argument can be 0 or bitwise OR of one or more of the
following file creation flags:
.IP \[bu] 2
\f[C]PMEM_FILE_CREATE\f[] \- Create the named file if it does not exist.
\f[C]len\f[] must be non\-zero and specifies the size of the file to be
created.
\f[C]mode\f[] has the same meaning as for \f[B]open\f[](2) and specifies
the mode to use in case a new file is created.
If neither \f[C]PMEM_FILE_CREATE\f[] nor \f[C]PMEM_FILE_TMPFILE\f[] is
specified, then \f[C]mode\f[] is ignored.
.IP \[bu] 2
\f[C]PMEM_FILE_EXCL\f[] \- Same meaning as \f[C]O_EXCL\f[] on
\f[B]open\f[](2) \- Ensure that this call creates the file.
If this flag is specified in conjunction with \f[C]PMEM_FILE_CREATE\f[],
and pathname already exists, then \f[C]pmem_map_file()\f[] will fail.
.IP \[bu] 2
\f[C]PMEM_FILE_TMPFILE\f[] \- Same meaning as \f[C]O_TMPFILE\f[] on
\f[B]open\f[](2).
Create a mapping for an unnamed temporary file.
\f[C]PMEM_FILE_CREATE\f[] and \f[C]len\f[] must be specified and
\f[C]path\f[] must be an existing directory name.
.IP \[bu] 2
\f[C]PMEM_FILE_SPARSE\f[] \- When creating a file, create a sparse
(holey) file instead of calling \f[B]posix_fallocate\f[](2).
Valid only if specified in conjunction with \f[C]PMEM_FILE_CREATE\f[] or
\f[C]PMEM_FILE_TMPFILE\f[], otherwise ignored.
.PP
If creation flags are not supplied, then \f[C]pmem_map_file()\f[]
creates a mapping for an existing file.
In such case, \f[C]len\f[] should be zero.
The entire file is mapped to memory; its length is used as the length of
the mapping and returned via \f[C]mapped_lenp\f[].
.PP
To delete mappings created with \f[C]pmem_map_file()\f[], use
\f[C]pmem_unmap()\f[].
.IP
.nf
\f[C]
int\ pmem_unmap(void\ *addr,\ size_t\ len);
\f[]
.fi
.PP
The \f[C]pmem_unmap()\f[] function deletes all the mappings for the
specified address range, and causes further references to addresses
within the range to generate invalid memory references.
It will use the address specified by the parameter \f[C]addr\f[], where
\f[C]addr\f[] must be a previously mapped region.
\f[C]pmem_unmap()\f[] will delete the mappings using the
\f[B]munmap\f[](2), On success, \f[C]pmem_unmap()\f[] returns zero.
On error, \-1 is returned, and \f[C]errno\f[] is set appropriately.
.SS PARTIAL FLUSHING OPERATIONS
.PP
The functions in this section provide access to the stages of flushing
to persistence, for the less common cases where an application needs
more control of the flushing operations than the \f[C]pmem_persist()\f[]
function described above.
.IP
.nf
\f[C]
void\ pmem_flush(const\ void\ *addr,\ size_t\ len);
\f[]
.fi
.IP
.nf
\f[C]
void\ pmem_drain(void);
\f[]
.fi
.PP
These functions provide partial versions of the \f[C]pmem_persist()\f[]
function described above.
\f[C]pmem_persist()\f[] can be thought of as this:
.IP
.nf
\f[C]
void\ pmem_persist(const\ void\ *addr,\ size_t\ len)
{
\ \ /*\ flush\ the\ processor\ caches\ */
\ \ pmem_flush(addr,\ len);

\ \ /*\ wait\ for\ any\ pmem\ stores\ to\ drain\ from\ HW\ buffers\ */
\ \ pmem_drain();
}
\f[]
.fi
.PP
These functions allow advanced programs to create their own variations
of \f[C]pmem_persist()\f[].
For example, a program that needs to flush several discontiguous ranges
can call \f[C]pmem_flush()\f[] for each range and then follow up by
calling \f[C]pmem_drain()\f[] once.
.IP
.nf
\f[C]
int\ pmem_has_hw_drain(void);
\f[]
.fi
.PP
The \f[C]pmem_has_hw_drain()\f[] function returns true if the machine
supports an explicit \f[I]hardware drain\f[] instruction for persistent
memory.
On Intel processors with persistent memory, stores to persistent memory
are considered persistent once they are flushed from the CPU caches, so
this function always returns false.
Despite that, programs using \f[C]pmem_flush()\f[]to flush ranges of
memory should still follow up by calling \f[C]pmem_drain()\f[] once to
ensure the flushes are complete.
As mentioned above, \f[C]pmem_persist()\f[] handles calling both
\f[C]pmem_flush()\f[] and \f[C]pmem_drain()\f[].
.SS COPYING TO PERSISTENT MEMORY
.PP
The functions in this section provide optimized copying to persistent
memory.
.IP
.nf
\f[C]
void\ *pmem_memmove_persist(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len);
\f[]
.fi
.IP
.nf
\f[C]
void\ *pmem_memcpy_persist(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len);
\f[]
.fi
.IP
.nf
\f[C]
void\ *pmem_memset_persist(void\ *pmemdest,\ int\ c,\ size_t\ len);
\f[]
.fi
.PP
The \f[C]pmem_memmove_persist()\f[], \f[C]pmem_memcpy_persist()\f[], and
\f[C]pmem_memset_persist()\f[], functions provide the same memory
copying as their namesakes \f[B]memmove\f[](3) \f[B]memcpy\f[](3), and
\f[B]memset\f[](3), and ensure that the result has been flushed to
persistence before returning.
For example, the following code is functionally equivalent to
\f[C]pmem_memmove_persist()\f[]:
.IP
.nf
\f[C]
void\ *
pmem_memmove_persist(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len)
{
\ \ void\ *retval\ =\ memmove(pmemdest,\ src,\ len);
\ \ pmem_persist(pmemdest,\ len);

return\ retval;
}
\f[]
.fi
.PP
Calling \f[C]pmem_memmove_persist()\f[] may out\-perform the above code,
however, since the \f[B]libpmem\f[] implementation may take advantage of
the fact that \f[C]pmemdest\f[] is persistent memory and use
instructions such as \f[C]non\-temporal\f[] stores to avoid the need to
flush processor caches.
.RS
.PP
WARNING: Using these functions where \f[C]pmem_is_pmem()\f[] returns
false may not do anything useful.
Use the normal libc functions in that case.
.RE
.IP
.nf
\f[C]
void\ *pmem_memmove_nodrain(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len);
\f[]
.fi
.IP
.nf
\f[C]
void\ *pmem_memcpy_nodrain(void\ *pmemdest,\ const\ void\ *src,\ size_t\ len);
\f[]
.fi
.IP
.nf
\f[C]
void\ *pmem_memset_nodrain(void\ *pmemdest,\ int\ c,\ size_t\ len);
\f[]
.fi
.PP
The \f[C]pmem_memmove_nodrain()\f[], \f[C]pmem_memcpy_nodrain()\f[] and
\f[C]pmem_memset_nodrain()\f[] functions are similar to
\f[C]pmem_memmove_persist()\f[], \f[C]pmem_memcpy_persist()\f[], and
\f[C]pmem_memset_persist()\f[] described above, except they skip the
final \f[C]pmem_drain()\f[] step.
This allows applications to optimize cases where several ranges are
being copied to persistent memory, followed by a single call to
\f[C]pmem_drain()\f[].
The following example illustrates how these functions might be used to
avoid multiple calls to \f[C]pmem_drain()\f[] when copying several
ranges of memory to pmem:
.IP
.nf
\f[C]
/*\ \&...\ write\ several\ ranges\ to\ pmem\ \&...\ */
pmem_memcpy_nodrain(pmemdest1,\ src1,\ len1);
pmem_memcpy_nodrain(pmemdest2,\ src2,\ len2);
/*\ \&...\ */
/*\ wait\ for\ any\ pmem\ stores\ to\ drain\ from\ HW\ buffers\ */
pmem_drain();
\f[]
.fi
.RS
.PP
WARNING: Using \f[C]pmem_memmove_nodrain()\f[],
\f[C]pmem_memcpy_nodrain()\f[] or \f[C]pmem_memset_nodrain()\f[] on a
destination where \f[C]pmem_is_pmem()\f[] returns false may not do
anything useful.
.RE
.SS LIBRARY API VERSIONING
.PP
This section describes how the library API is versioned, allowing
applications to work with an evolving API.
.IP
.nf
\f[C]
const\ char\ *pmem_check_version(unsigned\ major_required,\ unsigned\ minor_required);
\f[]
.fi
.PP
The \f[C]pmem_check_version()\f[] function is used to see if the
installed \f[B]libpmem\f[] supports the version of the library API
required by an application.
The easiest way to do this is for the application to supply the
compile\-time version information, supplied by defines in
\f[C]<libpmem.h>\f[], like this:
.IP
.nf
\f[C]
reason\ =\ pmem_check_version(PMEM_MAJOR_VERSION,\ PMEM_MINOR_VERSION);
if\ (reason\ !=\ NULL)
{
\ \ /*\ version\ check\ failed,\ reason\ string\ tells\ you\ why\ */
}
\f[]
.fi
.PP
Any mismatch in the major version number is considered a failure, but a
library with a newer minor version number will pass this check since
increasing minor versions imply backwards compatibility.
.PP
An application can also check specifically for the existence of an
interface by checking for the version where that interface was
introduced.
These versions are documented in this man page as follows: unless
otherwise specified, all interfaces described here are available in
version 1.0 of the library.
Interfaces added after version 1.0 will contain the text \f[I]introduced
in version x.y\f[] in the section of this manual describing the feature.
.PP
When the version check performed by \f[C]pmem_check_version()\f[] is
successful, the return value is NULL.
Otherwise the return value is a static string describing the reason for
failing the version check.
The string returned by \f[C]pmem_check_version()\f[] must not be
modified or freed.
.SS DEBUGGING AND ERROR HANDLING
.PP
Two versions of \f[B]libpmem\f[] are typically available on a
development system.
The normal version, accessed when a program is linked using the
\f[C]\-lpmem\f[] option, is optimized for performance.
That version skips checks that impact performance and never logs any
trace information or performs any run\-time assertions.
If an error is detected during the call to \f[B]libpmem\f[] function, an
application may retrieve an error message describing the reason of
failure using the following function:
.IP
.nf
\f[C]
const\ char\ *pmem_errormsg(void);
\f[]
.fi
.PP
The \f[C]pmem_errormsg()\f[] function returns a pointer to a static
buffer containing the last error message logged for current thread.
The error message may include description of the corresponding error
code (if \f[C]errno\f[] was set), as returned by \f[B]strerror\f[](3).
The error message buffer is thread\-local; errors encountered in one
thread do not affect its value in other threads.
The buffer is never cleared by any library function; its content is
significant only when the return value of the immediately preceding call
to \f[B]libpmem\f[] function indicated an error, or if \f[C]errno\f[]
was set.
The application must not modify or free the error message string, but it
may be modified by subsequent calls to other library functions.
.PP
A second version of \f[B]libpmem\f[], accessed when a program uses the
libraries under \f[B]/usr/lib/nvml_debug\f[], contains run\-time
assertions and trace points.
The typical way to access the debug version is to set the environment
variable \f[C]LD_LIBRARY_PATH\f[] to \f[B]/usr/lib/nvml_debug\f[] or
\f[B]/usr/lib64/nvml_debug\f[] depending on where the debug libraries
are installed on the system.
The trace points in the debug version of the library are enabled using
the environment variable \f[C]PMEM_LOG_LEVEL\f[], which can be set to
the following values:
.IP \[bu] 2
\f[B]0\f[] \- This is the default level when \f[C]PMEM_LOG_LEVEL\f[] is
not set.
No log messages are emitted at this level.
.IP \[bu] 2
\f[B]1\f[] \- Additional details on any errors detected are logged (in
addition to returning the \f[C]errno\f[]\-based errors as usual).
The same information may be retrieved using \f[C]pmem_errormsg\f[].
.IP \[bu] 2
\f[B]2\f[] \- A trace of basic operations is logged.
.IP \[bu] 2
\f[B]3\f[] \- This level enables a very verbose amount of function call
tracing in the library.
.IP \[bu] 2
\f[B]4\f[] \- This level enables voluminous and fairly obscure tracing
information that is likely only useful to the \f[B]libpmem\f[]
developers.
.PP
The environment variable \f[C]PMEM_LOG_FILE\f[] specifies a file name
where all logging information should be written.
If the last character in the name is “\-”, the PID of the current
process will be appended to the file name when the log file is created.
If \f[C]PMEM_LOG_FILE\f[] is not set, the logging output goes to stderr.
.PP
Setting the environment variable \f[C]PMEM_LOG_LEVEL\f[] has no effect
on the non\-debug version of \f[B]libpmem\f[].
.SS ENVIRONMENT VARIABLES
.PP
\f[B]libpmem\f[] can change its default behavior based on the following
environment variables.
These are largely intended for testing and are not normally required.
.IP \[bu] 2
\f[C]PMEM_IS_PMEM_FORCE\f[]=val
.PP
If \f[C]val\f[] is 0 (zero), then \f[C]pmem_is_pmem()\f[] will always
return false.
Setting \f[C]val\f[] to 1 causes \f[C]pmem_is_pmem()\f[] to always
return true.
This variable is mostly used for testing but can be used to force pmem
behavior on a system where a range of pmem is not detectable as pmem for
some reason.
.RS
.PP
NOTE: Unlike the other variables, the value of
\f[C]PMEM_IS_PMEM_FORCE\f[] is not queried (and cached) at the library
initialization time, but on the first call to \f[C]pmem_is_pmem()\f[]
function.
It means that in case of \f[B]libpmemlog\f[], \f[B]libpmemblk\f[], and
\f[B]libpmemobj\f[] libraries, it may still be set or modified by the
program until the first attempt to create or open the persistent memory
pool.
.RE
.IP \[bu] 2
\f[C]PMEM_NO_CLWB\f[]=1
.PP
Setting this environment variable to 1 forces \f[B]libpmem\f[] to never
issue the \f[C]CLWB\f[] instruction on Intel hardware, falling back to
other cache flush instructions instead (\f[C]CLFLUSHOPT\f[] or
\f[C]CLFLUSH\f[] on Intel hardware).
Without this environment variable, \f[B]libpmem\f[] will always use the
\f[C]CLWB\f[] instruction for flushing processor caches on platforms
that support the instruction.
This variable is intended for use during library testing but may be
required for some rare cases where using \f[C]CLWB\f[] has a negative
impact on performance.
.IP \[bu] 2
\f[C]PMEM_NO_CLFLUSHOPT\f[]=1
.PP
Setting this environment variable to 1 forces \f[B]libpmem\f[] to never
issue the \f[C]CLFLUSHOPT\f[] instruction on Intel hardware, falling
back to the \f[C]CLFLUSH\f[] instructions instead.
Without this environment variable, \f[B]libpmem\f[] will always use the
\f[C]CLFLUSHOPT\f[] instruction for flushing processor caches on
platforms that support the instruction, but where \f[C]CLWB\f[] is not
available.
This variable is intended for use during library testing.
.IP \[bu] 2
\f[C]PMEM_NO_MOVNT\f[]=1
.PP
Setting this environment variable to 1 forces \f[B]libpmem\f[] to never
use the \f[I]non\-temporal\f[] move instructions on Intel hardware.
Without this environment variable, \f[B]libpmem\f[] will use the
non\-temporal instructions for copying larger ranges to persistent
memory on platforms that support the instructions.
This variable is intended for use during library testing.
.IP \[bu] 2
\f[C]PMEM_MOVNT_THRESHOLD\f[]=val
.PP
This environment variable allows overriding the minimal length of
\f[C]pmem_memcpy_*()\f[], \f[C]pmem_memmove_*()\f[] or
\f[C]pmem_memset_*()\f[] operations, for which \f[B]libpmem\f[] uses
\f[I]non\-temporal\f[] move instructions.
Setting this environment variable to 0 forces \f[B]libpmem\f[] to always
use the \f[I]non\-temporal\f[] move instructions if available.
It has no effect if \f[C]PMEM_NO_MOVNT\f[] variable is set to 1.
This variable is intended for use during library testing.
.IP \[bu] 2
\f[C]PMEM_MMAP_HINT\f[]=val
.PP
This environment variable allows overriding the hint address used by
\f[C]pmem_map_file()\f[].
If set, it also disables mapping address randomization.
This variable is intended for use during library testing and debugging.
Setting it to some fairly large value (i.e.
0x10000000000) will very likely result in mapping the file at the
specified address (if not used) or at the first unused region above
given address, without adding any random offset.

ChangeLog

Thu Jun 23 2016 Krzysztof Czurylo <krzysztof.czurylo@intel.com>

	* Version 1.1
	This NVML release introduces a new version of libpmemobj pool layout.
	Internal undo log structure has been modified to improve performance
	of pmemobj transactions.  Memory pools created with older versions
	of the libpmemobj library must be converted to the new format using
	"pmempool convert" command.  See pmempool-convert(1) for details.

	A new "libpmempool" library is available, providing support for
	off-line pool management and diagnostics.  Initially it provides only
	"check" and "repair" operations for log and blk memory pools, and for
	BTT devices.

	Other changes:
	- pmem: deprecate PCOMMIT
	- blk: match BTT Flog initialization with Linux NVDIMM BTT
	- pmem: defer pmem_is_pmem() initialization (#158)
	- obj: add TOID_TYPEOF macro

	Bug fixes:
	- doc: update description of valid file size units (#133)
	- pmempool: fix --version short option in man page (#135)
	- pmempool: print usage when running rm without arg (#136)
	- cpp: clarify polymorphism in persistent_ptr (#150)
	- obj: let the before flag be any non-zero value (#151)
	- obj: fix compare array pptr to nullptr (#152)
	- obj: cpp pool.get_root() fix (#156)
	- log/blk: set errno if replica section is specified (#161)
	- cpp: change exception message	(#163)
	- doc: remove duplicated words in man page (#164)
	- common: always append EXTRA_CFLAGS after our CFLAGS

	Experimental features:
	- Implementation of C++ bindings for libpmempobj is complete.
	  Web-based documentation for C++ API is available on http://pmem.io.
	  Note that C++ API is still considered experimental.  Do not use it
	  in production environments.
	- Porting NVML to Windows is in progress.  There are MS Visual Studio
	  solution/projects available, allowing to compile libpmem, libpmemlog,
	  libpmemblk and libpmemobj on Windows, but the libraries are not fully
	  functional and most of the test are not enabled yet.

Thu Apr 07 2016 Krzysztof Czurylo <krzysztof.czurylo@intel.com>

	* Version 1.0
	The API of six libraries (libpmem, libpmemblk, libpmemlog,
	libpmemobj, libvmem, libvmmalloc) is complete and stable.
	The on-media layout of persistent memory pools will be maintained
	from this point, and if changed it will be backward compatible.
	Man pages are all complete.
	This release has been validated to "Production quality".

	For the purpose of new features planned for next releases of NVML
	there have been some API modifications made:
		- pmem: pmem_map replaced with pmem_map_file
		- log/blk: 'off_t' substituted with 'long long'
		- obj: type numbers extended to 64-bit
		- obj: new entry points and macros added:
		  pmemobj_tx_errno, pmemobj_tx_lock, pmemobj_mutex_timedlock,
		  TX_ADD_DIRECT, TX_ADD_FIELD_DIRECT, TX_SET_DIRECT

	Other key changes since version 0.4 include:
		- common: updated/fixed installation scripts
		- common: eliminated dependency on libuuid
		- pmem: CPU features/ISA detection using CPUID
		- obj: improved error handling
		- obj: atomic allocation fails if constructor returns error
		- obj: multiple performance optimizations
		- obj: object store refactoring
		- obj: additional examples and benchmarks

	This release also introduces a prototype implementation
	of C++ bindings for libpmemobj.  Note that C++ API is still
	experimental and should not be used in production environments.

Fri Dec 04 2015 Krzysztof Czurylo <krzysztof.czurylo@intel.com>

	* Version 0.4
	This NVML version primarily focuses on improving code quality
	and reliability.  In addition to a couple of bug fixes,
	the changes include:
		- benchmarks for libpmemobj, libpmemblk and libvmem
		- additional pmemobj tests and examples
		- pool mapping address randomization
		- added pmempool "rm" command
		- eliminated libpmem dependency on libpthread
		- enabled extra warnings
		- minor performance improvements
	Man pages are all complete.
	This release is considered "Beta quality" by the team, having
	been thoroughly validated, including significant performance analysis.
	The pmempool command does not yet support "check" and "repair"
	operations for pmemobj type pools.

Sun Sep 13 2015 Andy Rudoff <andy.rudoff@intel.com>

	* Version 0.3
	NVML is now feature complete, adding support for:
		- pool sets
		- pmemobj local replication (active/passive)
		- experimental valgrind support
		- pmempool support for all pool types
	Man pages are all complete.
	This release is considered "Alpha quality" by the team, having
	gone through significant validation but only some performance
	analysis at this point.

Tue Jun 30 2015 Andy Rudoff <andy.rudoff@intel.com>

	* Version 0.2
	NVML now consists of six libraries:
		- libpmem                            (basic flushing, etc)
		- libpmemblk, libpmemlog, libpmemobj (transactions)
		- libvmem, libvmmalloc               (volatile use of pmem)
	The "pmempool" command is available for managing pmem files.
	Man pages for all the above are complete.
	The only things documented in man pages but not implemented are:
		- pmem sets (ability to spread a pool over a set of files)
		- replication (coming for libpmemobj)
	The pmempool command does not yet support pmemobj type pools.

Thu Sep 11 2014 Andy Rudoff <andy.rudoff@intel.com>

	* Version 0.1
	Initial development done in 0.1 builds


src/libpmem/libpmem.c
/*
 * Copyright 2014-2016, Intel Corporation
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 *     * Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * libpmem.c -- pmem entry points for libpmem
 */

#include <stdio.h>
#include <stdint.h>

#include "libpmem.h"

#include "pmem.h"
#include "mmap.h"
#include "out.h"

/*
 * libpmem_init -- load-time initialization for libpmem
 *
 * Called automatically by the run-time loader.
 */
ATTR_CONSTRUCTOR
void
libpmem_init(void)
{
	util_init();
	out_init(PMEM_LOG_PREFIX, PMEM_LOG_LEVEL_VAR, PMEM_LOG_FILE_VAR,
			PMEM_MAJOR_VERSION, PMEM_MINOR_VERSION);
	LOG(3, NULL);
	mmap_init();
	pmem_init();
}

/*
 * libpmem_fini -- libpmem cleanup routine
 *
 * Called automatically when the process terminates.
 */
ATTR_DESTRUCTOR
void
libpmem_fini(void)
{
	LOG(3, NULL);

	out_fini();
}

/*
 * pmem_check_version -- see if library meets application version requirements
 */
const char *
pmem_check_version(unsigned major_required, unsigned minor_required)
{
	LOG(3, "major_required %u minor_required %u",
			major_required, minor_required);

	if (major_required != PMEM_MAJOR_VERSION) {
		ERR("libpmem major version mismatch (need %u, found %u)",
			major_required, PMEM_MAJOR_VERSION);
		return out_get_errormsg();
	}

	if (minor_required > PMEM_MINOR_VERSION) {
		ERR("libpmem minor version mismatch (need %u, found %u)",
			minor_required, PMEM_MINOR_VERSION);
		return out_get_errormsg();
	}

	return NULL;
}

/*
 * pmem_errormsg -- return last error message
 */
const char *
pmem_errormsg(void)
{
	return out_get_errormsg();
}
